### PR TITLE
Denic version3.0

### PR DESCRIPTION
**Pull request by paullojorgge**

### CODE REVIEW

File: t/Net/DRI/Protocol/RRI.t [hunk 1]
```diff
@@ -48,65 +48,94 @@ $R2 = $E1 . '<tr:transaction><tr:stid>' . $TRID . '</tr:stid><tr:result>success<
...
 $c->cc('DE');
-$c->voice('+49.123456');
-$c->fax('+49.123457');
 $c->email('email@denic.de');
-$c->sip('sip:benutzer@denic.de');
-$c->remarks('Interesting remark');
-$c->disclose({voice =>1});
 
-# Contact create
+# Contact create (based on: https://member.secure.denic.de/index.php?eID=dumpFile&t=f&f=7279&token=c5ea0100484bd1c67f535dc0fef2a7d6e318db53)
```
Comment (thedarkwinter) [inline t/Net/DRI/Protocol/RRI.t]: This is not a public link ?

File: lib/Net/DRI/Protocol/RRI/Domain.pm [hunk 2]
```diff
@@ -541,35 +542,31 @@ sub trade
...
  my $mes = $rri->message();
  my %ns = map { $_ => $mes->ns->{$_}->[0] } qw(domain dnsentry xsi);
- my @d = build_command($mes, 'migrate-descr', $domain, undef, \%ns);
-
- ## Contacts, Holder is required
- push @d,build_contact($rd->{contact}) if Net::DRI::Util::has_contact($rd);
-
+ my $hash = exists($rd->{'authinfohash'}) ? { hash => $rd->{'authinfohash'}} : undef;
+ $hash->{'expire'} = $rd->{'authinfoexpire'} if ($hash && exists($rd->{'authinfoexpire'}));
+ my $cmd = ($hash) ? 'createAuthInfo1' : 'createAuthInfo2';
```
Comment (thedarkwinter) [inline lib/Net/DRI/Protocol/RRI/Domain.pm]: What was the thinking behind adding a new function etc for createAuthInfo2 when it works via createAuthInfo1? I haven't looked too deep into it, but having 2 authcodes is not new, so I am trying to work out what has changed that has made you split this into two functions?

File: t/Net/DRI/Protocol/RRI.t [hunk 3]
```diff
@@ -48,65 +48,94 @@ $R2 = $E1 . '<tr:transaction><tr:stid>' . $TRID . '</tr:stid><tr:result>success<
...
+$rc = $dri->contact_info($dri->local_object('contact')->srid('DENIC-99995-GENREQ'));
 isa_ok($rc, 'Net::DRI::Protocol::ResultStatus');
-is(defined($rc) && $rc->is_success(), 1, 'Contact successfully checked');
-is_string($R1, '<?xml version="1.0" encoding="UTF-8" standalone="yes"?><registry-request xmlns="http://registry.denic.de/global/1.0" xmlns:contact="http://registry.denic.de/contact/1.0"><contact:check><contact:handle>DENIC-99990-10240-BSP</contact:handle></contact:check></registry-request>', 'Check Contact XML correct');
-is($dri->get_info('exist', 'contact', 'DENIC-99990-10240-BSP'), 1, 'Contact exists');
+is($rc->is_success(), 1, 'Contact successfully queried (version 3.0)');
+is_string($R1, '<?xml version="1.0" encoding="UTF-8" standalone="yes"?><registry-request xmlns="http://registry.denic.de/global/3.0" xmlns:contact="http://registry.denic.de/contact/3.0"><contact:info><contact:handle>DENIC-99995-GENREQ</contact:handle></contact:info></registry-request>', 'Query Contact XML correct (version 3.0)');
+$c = $dri->get_info('self', 'contact', 'DENIC-99995-GENREQ');
+isa_ok($c, 'Net::DRI::Data::Contact::DENIC');
+is($c->type() . '|' . $c->uri_template(),
```
Comment (thedarkwinter) [inline t/Net/DRI/Protocol/RRI.t]: This is an odd test? Why are you testing these two variables concatenated with | instead of testing separately?

Comment (thedarkwinter) [reply t/Net/DRI/Protocol/RRI.t]: Oh, isee some older tests were done like that...

File: lib/Net/DRI/DRD/DENIC.pm [hunk 4]
```diff
@@ -171,5 +173,19 @@ sub domain_delete_authinfo
...
 
+sub domain_restore
+{
+ my ($self, $reg, $dom, $rd) = @_;
+ return $reg->process('domain', 'restore', [$dom, $rd]);
+}
+
+# no need to create a new object (not even listed in their XSD only on their technical documentation)
+# RegAcc INFO request you can query your own public registrar contact details or those of others
+sub regacc_info
```
Comment (thedarkwinter) [inline lib/Net/DRI/DRD/DENIC.pm]: I wonder if its worth using registrar_info (like eurid, afilias, and I think ZA?). So in the DENIC module you can add the comment regacc, but the function itself would be named registrar_info to be the same as other registries?

File: t/Net/DRI/Protocol/RRI.t (hunk 1 above)
Comment (paullojorgge) [reply t/Net/DRI/Protocol/RRI.t]: Removed member link and substitute by something like: _(based on RRI v3.0 sample)_ instead

File: lib/Net/DRI/Protocol/RRI/Domain.pm (hunk 2 above)
Comment (paullojorgge) [reply lib/Net/DRI/Protocol/RRI/Domain.pm]: Hum... you are right. Didn't notice _$cmd_ on line 559 that handle 1 and 2 depending of the existence of a hash. My bad!

Restored to how it was and tests updated to reflect the changes:
- https://github.com/thedarkwinter/Net-DRI/pull/343/commits/c1c34ff8e3bde935c265f3f4fe4b490e166e2822

File: t/Net/DRI/Protocol/RRI.t (hunk 3 above)
Comment (paullojorgge) [reply t/Net/DRI/Protocol/RRI.t]: Yep. Kept to how it was before :)

File: lib/Net/DRI/DRD/DENIC.pm (hunk 4 above)
Comment (paullojorgge) [reply lib/Net/DRI/DRD/DENIC.pm]: Why not. Do you agree with this approach?
- https://github.com/thedarkwinter/Net-DRI/pull/343/commits/61fcc7e61d31e3d5cf59b0904dc7d9c5f4349c51

Comment (thedarkwinter) [reply lib/Net/DRI/DRD/DENIC.pm]: yip thanks fine

**Pull request merged**